### ISSUE TITLE
As a moderator, I want to create and edit related links in wagtail so that I can associate them with projects and people.

### ISSUE THREAD

**Issue by thatbudakguy** [Princeton-CDH/cdh-web#181]
# testing notes
fyi - "related links" were previously called "resources"; we've renamed them to be a bit more clear. they allow you to add a particular link to a person or project, which will show up on the corresponding page or profile.

the "website" related link is special - if you add a website to a person, clicking on them on a list page will take you to that website (unless they already have a profile on the CDH site). we do this often for faculty, who have profile pages on departmental websites that we want to link to instead of making them a profile on our site. if you add a website to a project, the website will be displayed in a special way on the project's page, showing its URL.

### projects
1. go to the admin site and edit any project. scroll down to the bottom until you find the "related links" section.
2. add a new related link of any type. if the project doesn't have a website link, add a website link too.
3. visit the project's page by clicking "preview". you should see the link you added, as well as the website link, and the website link should show the URL you entered.
### people
1. go to the admin site and edit any non-staff CDH person. scroll down to the bottom until you find the "related links" section.
2. add a website link for this person.
3. find the person on whichever person list page they will show up (e.g. students) and click them - you should go to the website URL you added.
4. go back and edit any CDH staff member, repeating step 1 above.
5. add any kind of related link for this person you like, including a website link.
6. visit the person's profile by clicking them on the staff list page. this should take you to their profile, _not_ the website URL you added (we only use that URL if you don't have a profile on our site).
7. you should see any other related links you added on the person's profile, shown on the left.

# dev notes
This is ready to be worked on once #268 is merged.

## background
**n.b.**
You will see the term "Resource" used throughout the codebase, but the title of this issue uses the term "Related Link"; this is the term we want to use going forward because it's more descriptive. Thus related links for people will be renamed from `PersonResource` to `PersonRelatedLink`, etc.

We currently allow associating arbitrary URLs (related links) with two models in cdh-web: `Person` and `Project`. This is used, for example, to display a list of relevant links on a person's profile page:

<img width="554" alt="Screen Shot 2021-01-08 at 11 08 22 AM" src="https://user-images.githubusercontent.com/4924494/104036860-efd17c00-51a1-11eb-93b1-886793eefde0.png">

More importantly, it lets us link to profile pages hosted on external sites, such as departmental websites. We often do this for faculty who already have a profile page elsewhere; see the link below for Marina which goes to `nes.princeton.edu`. This functionality is powered by creating a `PersonResource` (a related link) with a `ResourceType` of `website`. When a `Person` does not have a `ProfilePage` internally (on the cdh site), this URL is used instead.

<img width="1002" alt="Screen Shot 2021-01-08 at 11 10 41 AM" src="https://user-images.githubusercontent.com/4924494/104037029-2a3b1900-51a2-11eb-86d4-25279819229b.png">

Part of the utility of these related links is that you can edit them directly ("inline") when changing a `Person` or `Project` in the admin backend, rather than needing to individually create links one-by-one and then associate them with the `Person` or `Project` (see below screenshot). We want to preserve this functionality, but Wagtail has a different way of supporting it, which forms the basis for this issue.

<img width="1186" alt="Screen Shot 2021-01-08 at 11 17 14 AM" src="https://user-images.githubusercontent.com/4924494/104038415-16dc7d80-51a3-11eb-9cc4-12c33c376568.png">

## implementation
the wagtail [docs on inlines](https://docs.wagtail.io/en/v2.5/reference/pages/panels.html#inline-panels-and-model-clusters) specifically address our use case, and we can use their example with only slight modifications. the idea is that we have an abstract model `RelatedLink` that describes relating a link to _anything_, and then concrete models extending it that describe relating links to _particular_ things (`Person` and `Project`, in our case).

the first and simplest thing we want to do is to rename the `ResourceType` model to `RelatedLinkType` for more clarity, since the name "resource" doesn't mean very much. take a look at `cdhweb/resources/migrations/0008_userresource_to_personresource.py` for an example of a migration that does this with `migrations.RenameModel`. you can rename the model in `resources/models.py` and save the file, then create the migration manually using `python manage.py makemigrations resources --empty` and edit the resulting file to add the `RenameModel` command. make sure to give the generated migration file a descriptive name.

next we will want to copy the wagtail docs `RelatedLink` example into `resources/models.py`. we just need to change the `title` field to a `ForeignKey` to `RelatedLinkType`, since that's where that information is stored. we also call the url field `url` instead of `link_external`, and don't allow it to be blank. so it would look like:

```py
class RelatedLink(models.Model):
    type = models.ForeignKey(RelatedLinkType, on_delete=models.CASCADE)
    url = models.URLField()

    panels = [
        FieldPanel('type'),
        FieldPanel('url'),
    ]

    class Meta:
        abstract = True
```

you can see how this model now abstracts away the repeated parts of `PersonResource` and `ProjectResource` (`type` and `url`). create another migration using `python manage.py makemigrations resources`; django should populate it correctly automatically.

now we just need to create concrete versions of `RelatedLink` for `Person` and `Project` - `PersonRelatedLink` and `ProjectRelatedLink`. here's an example for `Person`:

```py
class PersonRelatedLink(Orderable, RelatedLink):
    person = ParentalKey(Person, on_delete=models.CASCADE, related_name="related_links")
```

instead of storing them in `resources`, let's put them in the apps to which their "related" model belongs: `people/models.py` for `PersonRelatedLink` and `projects/models.py` for `ProjectRelatedLink`. then we can again create migrations for each of those apps, which should be generated automatically.

finally, we need to make these related links inline-editable on their respective models. in the wagtail example, that's `BookPage` - for us it will be `Person` and `Project`. for `Person`, all we need to do is add another panel to the available panels in the admin, e.g.:

```py
class Person(ClusterableModel):
  # ...

  content_panels = Page.content_panels + [
    #....
    InlinePanel("related_links", heading="Related Links"),
  ]

```

`Project` isn't yet a wagtail `Page` model, however, so we can't do the same there until #182 is completed.

there's not much to test here, since all the logic is managed by Django in the form of database fields/models. the only thing to update is to rename/update the test for `ResourceType` in `resources/tests/test_models.py`, since it got renamed. you could also look for lingering instances of "resource" to rename across the codebase.

if you want to check out how things looks, fire up a dev server and visit `localhost:8000/cms`. if you select a person in the wagtail admin, you should see new fields for adding and removing related links.

that's everything this issue covers - the actual migration of `PersonResource`s and `ProjectResource`s into `PersonRelatedLink`s and `ProjectRelatedLink`s will be #249.

## checklist
- [x] migration to rename `ResourceType` to `RelatedLinkType`
- [x] create `RelatedLink` model and migration
- [x] create `PersonRelatedLink` model and migration
- [x] create `ProjectRelatedLink` model and migration
- [x] make `PersonRelatedLink` inline-editable on `Person`
- [x] make `ProjectRelatedLink` inline-editable on `Project`
- [x] update tests & check for any remaining references to "resource"
- [x] verify things look right in wagtail admin

**Comment by rlskoeser** [comment]
FWIW, django may be able to detect the model rename and generate the migration for you. (IIRC it asks if you if you did a rename)

**Comment by kmcelwee** [comment]
Testing notes:
- [ ] Go into the admin interface and add and remove a related link from a *project* and make sure the changes are reflected in the page.
- [ ] Go into the admin interface and add and remove a related link from a *profile* and make sure the changes are reflected in the page.

**Comment by thatbudakguy** [comment]
can't test this for projects yet since they're not migrated to wagtail.

**Comment by manawinters** [comment]
Under Projects, I tried adding links to Philly Wireless Project and Indigenous DH Working Group and "received the chicken"

Under People, I successfully added website links to both non-staff (Ipsita Dey) and staff (Sierra) members.

**Comment by manawinters** [comment]
- The link to Philly Wireless Project now works. 

- The Working Groups still do not show up.

- Ipsita Dey's email address does not appear on her page of the [test site](https://cdh-test-web.princeton.edu/pages/people/ipsita-dey/) like it does in the current [production site](https://cdh.princeton.edu/people/ipsita-dey/). Also, when I add a website link to Ipsita, I am not directed there when I click on her on the list page; instead her [website](https://cdh-test-web.princeton.edu/pages/people/ipsita-dey/) appears to the left of her photo.

**Comment by thatbudakguy** [comment]
working groups issue should be tested/fixed on #280.

@manawinters can you test and see if Ipsita's email appears now? if so, this is ready to close.
